File: macos_arm64/Sources/RimeoAgentMac/Models/DataStore.swift
import Foundation
import CryptoKit

// Overlay record for a user-editable playlist (Фаза 0 — плейлисты из iOS). Lives
// in rimo_data.json next to notes/exclusions. `master.db` is never written in v1;
// these overlays are merged into /api/data at read time. A pure-Rimeo playlist has
// `rekordbox_id == nil` (synthetic path `rmo:<rimeo_id>`); an edited Rekordbox
// playlist carries the real `rekordbox_id` and `dirty == true` (overlay overrides
// the parsed membership by id, never by path — same-named playlists collide).
struct PlaylistOverlay: Codable {
    var rimeo_id:         String   = ""
    var rekordbox_id:     String?  = nil
    var name:             String   = ""
    var parent:           String?  = nil
    var is_folder:        Bool     = false
    var track_ids:        [String] = []
    var state:            String   = "pending"
    var dirty:            Bool     = true
    var content_hash:     String?  = nil
    var last_synced_hash: String?  = nil
    // Tombstone (full-CRUD delete). A deleted overlay hides the playlist from
    // /api/data: a pure-Rimeo one is simply not emitted; an edited Rekordbox one
    // is removed from the parsed list (master.db is still never written in v1).
    var deleted:          Bool     = false
}

// Resilient decode (finding-20): decode each key with `decodeIfPresent ?? default`
// so a rimo_data.json written before a field was added never throws keyNotFound
// (which load()'s `try?` would swallow → the whole store resets → agent unpairs).
// In an extension so the memberwise `init()` stays synthesized for construction.
extension PlaylistOverlay {
    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        self.init()
        rimeo_id         = try c.decodeIfPresent(String.self,   forKey: .rimeo_id)         ?? rimeo_id
        rekordbox_id     = try c.decodeIfPresent(String.self,   forKey: .rekordbox_id)     ?? rekordbox_id
        name             = try c.decodeIfPresent(String.self,   forKey: .name)             ?? name
        parent           = try c.decodeIfPresent(String.self,   forKey: .parent)           ?? parent
        is_folder        = try c.decodeIfPresent(Bool.self,     forKey: .is_folder)        ?? is_folder
        track_ids        = try c.decodeIfPresent([String].self, forKey: .track_ids)        ?? track_ids
        state            = try c.decodeIfPresent(String.self,   forKey: .state)            ?? state
        dirty            = try c.decodeIfPresent(Bool.self,     forKey: .dirty)            ?? dirty
        content_hash     = try c.decodeIfPresent(String.self,   forKey: .content_hash)     ?? content_hash
        last_synced_hash = try c.decodeIfPresent(String.self,   forKey: .last_synced_hash) ?? last_synced_hash
        deleted          = try c.decodeIfPresent(Bool.self,     forKey: .deleted)          ?? deleted
    }
}

// Cross-platform playlist content hash (finding-7/25). MUST stay byte-identical to
// the iOS / web / Windows implementations or overlays never self-clear: SHA-256,
// lowercase hex, of the ordered `track_ids` joined by "\n". Order is significant;
// ids are not case-normalized. Single source of truth for every mutation handler.
enum PlaylistHash {
    static func contentHash(_ trackIDs: [String]) -> String {
        let joined = trackIDs.joined(separator: "\n")
        let digest = SHA256.hash(data: Data(joined.utf8))
        return digest.map { String(format: "%02x", $0) }.joined()
    }
}

// Sync signature (Фаза 6). `content_hash` covers ONLY membership (SHA-256 of the
// ordered track_ids) — a rename, a move, a folder create or a delete leaves it
// byte-identical, so a content_hash-based pending check would silently declare
// those changes already-synced. The sync signature covers everything the writer
// can push into master.db: name, parent, folder-ness, tombstone AND membership.
// `last_synced_hash` stores the signature of the state that was actually written,
// so `last_synced_hash != syncSignature` is exactly "there is something to sync".
extension PlaylistOverlay {
    var syncSignature: String {
        let parts = [name,
                     parent ?? "",
                     is_folder ? "1" : "0",
                     deleted   ? "1" : "0",
                     track_ids.joined(separator: "\n")]
        // \u{1F} (unit separator) can't occur in a name/path/track id → no collision
        // between e.g. name="a", parent="b" and name="a\u{1F}b", parent="".
        return PlaylistHash.contentHash([parts.joined(separator: "\u{1F}")])
    }
}

// Persistent data stored in rimo_data.json
struct RimoData: Codable {
    var notes:             [String: String]    = [:]
    // Custom display names for Rekordbox play-history sessions (djmdHistory.ID →
    // user-chosen name). Rekordbox's own DB is read-only to us, so renames live
    // here and are applied to the parsed library on every read. Synced to all
    // clients (iOS + web) because both read names from this agent.
    var history_names:     [String: String]    = [:]
    var global_exclusions: [String]            = []
    var pairing_code:      String              = ""
    // Per-device LAN pre-shared key (M4). Strong, persistent (unlike the rotating
    // 5-char pairing_code). Authorises direct local-network requests without a
    // server JWT; emitted in the v2 QR as `secret` and sent back as `?lan_token=`.
    var lan_secret:        String              = ""
    var cloud_url:         String              = ""
    var cloud_user_id:     String?             = nil
    var cloud_token:       String              = ""
    var tunnel_url:        String              = ""
    var max_cache_gb:      Double              = 3.0
    var just_updated:      Bool               = false
    var pending_update_url: String            = ""
    var pending_update_tag: String            = ""
    // Silent auto-update: hourly check downloads the new build's zip to a staging
    // file in the background; it is applied (extract+replace+relaunch) on the next
    // launch. `staged_update_tag` non-empty = a staged build is ready to install.
    var staged_update_tag: String             = ""
    // Overlay-only playlist edits from iOS (Фаза 0). Merged into /api/data at read
    // time; never written back to Rekordbox's master.db in v1.
    var playlists:         [PlaylistOverlay]   = []
}

// Resilient decoding. The synthesized `Codable.init(from:)` calls `decode` (not
// `decodeIfPresent`) for every non-optional field and IGNORES the struct's
// default values — so an older `rimo_data.json` that predates a newly-added
// field would throw `keyNotFound`, get swallowed by `load()`'s `try?`, and reset
// the whole store to empty → the agent would silently unpair on every update
// that introduced a field. Decoding each key with `decodeIfPresent ?? default`
// makes schema evolution forward/backward compatible. (Defined in an extension
// so the memberwise `init()` used by `RimoData()` stays synthesized.)
extension RimoData {
    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        self.init()
        notes              = try c.decodeIfPresent([String: String].self, forKey: .notes)             ?? notes
        history_names      = try c.decodeIfPresent([String: String].self, forKey: .history_names)     ?? history_names
        global_exclusions  = try c.decodeIfPresent([String].self,         forKey: .global_exclusions) ?? global_exclusions
        pairing_code       = try c.decodeIfPresent(String.self,           forKey: .pairing_code)      ?? pairing_code
        lan_secret         = try c.decodeIfPresent(String.self,           forKey: .lan_secret)        ?? lan_secret
        cloud_url          = try c.decodeIfPresent(String.self,           forKey: .cloud_url)         ?? cloud_url
        cloud_user_id      = try c.decodeIfPresent(String.self,           forKey: .cloud_user_id)     ?? cloud_user_id
        cloud_token        = try c.decodeIfPresent(String.self,           forKey: .cloud_token)       ?? cloud_token
        tunnel_url         = try c.decodeIfPresent(String.self,           forKey: .tunnel_url)        ?? tunnel_url
        max_cache_gb       = try c.decodeIfPresent(Double.self,           forKey: .max_cache_gb)      ?? max_cache_gb
        just_updated       = try c.decodeIfPresent(Bool.self,             forKey: .just_updated)      ?? just_updated
        pending_update_url = try c.decodeIfPresent(String.self,           forKey: .pending_update_url) ?? pending_update_url
        pending_update_tag = try c.decodeIfPresent(String.self,           forKey: .pending_update_tag) ?? pending_update_tag
        staged_update_tag  = try c.decodeIfPresent(String.self,           forKey: .staged_update_tag) ?? staged_update_tag
        playlists          = try c.decodeIfPresent([PlaylistOverlay].self, forKey: .playlists)        ?? playlists
    }
}

final class DataStore {
    static let shared = DataStore()

    private let queue = DispatchQueue(label: "rimeo.datastore", qos: .utility)
    private var _data = RimoData()

    var data: RimoData {
        queue.sync { _data }
    }

    private init() {
        _data = load()
    }

    private func load() -> RimoData {
        guard let raw = try? Data(contentsOf: AppConfig.shared.dataFile),
              let decoded = try? JSONDecoder().decode(RimoData.self, from: raw)
        else { return RimoData() }
        return decoded
    }

    func save(_ data: RimoData) {
        queue.sync { _data = data }
        DispatchQueue.global(qos: .utility).async {
            if let raw = try? JSONEncoder().encode(data) {
                // Atomic: a crash/shutdown mid-write must not leave a truncated
                // JSON that fails to decode and resets the store (→ unpair).
                try? raw.write(to: AppConfig.shared.dataFile, options: .atomic)
            }
        }
    }

    func update(_ block: (inout RimoData) -> Void) {
        var copy = data
        block(&copy)
        save(copy)
    }

    // Overlays that still differ from what was last written into master.db
    // (Фаза 6). Single source of truth for both the Sync engine and the
    // `pending_sync` counter in /api/data — they must never disagree, or the iOS
    // button offers a sync that turns into a no-op (or hides one that is needed).
    // A tombstone for a playlist that never reached the DB has nothing to delete
    // there, so it is not pending (it is dropped from the overlay list on its own).
    // `last_synced_hash == nil` ⇒ never synced ⇒ pending.
    func pendingSyncOverlays() -> [PlaylistOverlay] {
        data.playlists.filter { ov in
            if ov.deleted && (ov.rekordbox_id ?? "").isEmpty { return false }
            return ov.last_synced_hash != ov.syncSignature
        }
    }
}

File: windows_csharp/RimeoAgent/Models/DataStore.cs
using System.Security.Cryptography;
using System.Text;
using System.Text.Json;
using System.Text.Json.Serialization;
using RimeoAgent.Config;

namespace RimeoAgent.Models;

// Оверлей пользовательского плейлиста (Фаза 0 — плейлисты из iOS). Живёт в
// rimo_data.json рядом с notes/exclusions и подмешивается в /api/data на чтении;
// master.db на Windows пока НЕ пишется (см. capabilities.playlist_sync).
// Чисто-Rimeo плейлист: RekordboxId == null, синтетический путь "rmo:<rimeo_id>".
// Правка существующего Rekordbox-плейлиста: RekordboxId заполнен, Dirty == true —
// оверлей перекрывает разобранный состав ПО ID, никогда по пути (у одноимённых
// плейлистов путь совпадает и составы бы слились).
//
// System.Text.Json терпит отсутствующие ключи (поле получает значение по
// умолчанию), поэтому старый rimo_data.json без `playlists` десериализуется без
// исключения и агент не «распаривается» после обновления.
public sealed class PlaylistOverlay
{
    [JsonPropertyName("rimeo_id")]         public string  RimeoId        { get; set; } = "";
    [JsonPropertyName("rekordbox_id")]     public string? RekordboxId    { get; set; }
    [JsonPropertyName("name")]             public string  Name           { get; set; } = "";
    [JsonPropertyName("parent")]           public string? Parent         { get; set; }
    [JsonPropertyName("is_folder")]        public bool    IsFolder       { get; set; }
    [JsonPropertyName("track_ids")]        public List<string> TrackIds  { get; set; } = new();
    [JsonPropertyName("state")]            public string  State          { get; set; } = "pending";
    [JsonPropertyName("dirty")]            public bool    Dirty          { get; set; } = true;
    [JsonPropertyName("content_hash")]     public string? ContentHash    { get; set; }
    [JsonPropertyName("last_synced_hash")] public string? LastSyncedHash { get; set; }
    // Tombstone (полный CRUD-delete). Удалённый оверлей прячет плейлист из /api/data:
    // чисто-Rimeo просто не отдаётся, правка Rekordbox-плейлиста — выкидывается из
    // разобранного списка.
    [JsonPropertyName("deleted")]          public bool    Deleted        { get; set; }

    /// Подпись синка (Фаза 6). ContentHash покрывает ТОЛЬКО состав, поэтому
    /// переименование, перенос, создание папки и удаление оставляют его
    /// байт-в-байт прежним — проверка «что синкать» по ContentHash молча объявила
    /// бы эти правки уже записанными. Подпись покрывает всё, что писатель умеет
    /// протолкнуть в master.db: имя, родителя, папочность, tombstone И состав.
    [JsonIgnore]
    public string SyncSignature
    {
        get
        {
            // U+001F (unit separator) не встречается ни в имени, ни в пути, ни в id →
            // Escape записан ЯВНО: невидимый управляющий символ в исходнике рано или
            // поздно теряется при копировании, а хеш обязан совпадать с macOS.
            // нет коллизии между (name="a", parent="b") и (name="a\u001Fb", parent="").
            var parts = string.Join("\u001F", new[]
            {
                Name,
                Parent ?? "",
                IsFolder ? "1" : "0",
                Deleted  ? "1" : "0",
                string.Join("\n", TrackIds),
            });
            return PlaylistHash.ContentHash(new List<string> { parts });
        }
    }
}

/// Кросс-платформенный хеш состава плейлиста. ОБЯЗАН оставаться байт-в-байт
/// одинаковым с iOS / web / macOS, иначе оверлеи никогда не самоочистятся:
/// SHA-256 в нижнем регистре hex от упорядоченных track_ids, склеенных "\n".
/// Порядок значим; id не приводятся к одному регистру.
public static class PlaylistHash
{
    public static string ContentHash(IEnumerable<string> trackIds)
    {
        var joined = string.Join("\n", trackIds);
        var hash   = SHA256.HashData(Encoding.UTF8.GetBytes(joined));
        return Convert.ToHexString(hash).ToLowerInvariant();
    }
}

public sealed class RimoData
{
    [JsonPropertyName("notes")]             public Dictionary<string, string> Notes           { get; set; } = new();
    // Custom names for Rekordbox play-history sessions (djmdHistory.ID → name).
    // Applied to the parsed library on read; Rekordbox's DB stays untouched.
    [JsonPropertyName("history_names")]     public Dictionary<string, string> HistoryNames    { get; set; } = new();
    [JsonPropertyName("global_exclusions")] public List<string>               GlobalExclusions { get; set; } = new();
    [JsonPropertyName("pairing_code")]      public string                     PairingCode      { get; set; } = "";
    // Per-device LAN pre-shared key (M4). Persistent (unlike the rotating
    // PairingCode). Authorises direct local-network requests without a server JWT;
    // emitted in the v2 QR as `secret`, sent back as `?lan_token=`. Mirrors macOS.
    [JsonPropertyName("lan_secret")]        public string                     LanSecret        { get; set; } = "";
    [JsonPropertyName("cloud_url")]         public string                     CloudUrl         { get; set; } = "";
    [JsonPropertyName("cloud_user_id")]     public string?                    CloudUserId      { get; set; }
    [JsonPropertyName("cloud_token")]       public string                     CloudToken       { get; set; } = "";
    [JsonPropertyName("tunnel_url")]        public string                     TunnelUrl        { get; set; } = "";
    [JsonPropertyName("max_cache_gb")]      public double                     MaxCacheGb       { get; set; } = 3.0;
    // Silent auto-update: hourly check downloads the new build's zip to a staging
    // file in the background; it is applied (xcopy+restart) on the next launch.
    // Non-empty = a staged build is ready to install. Mirrors macOS.
    [JsonPropertyName("staged_update_tag")] public string                     StagedUpdateTag  { get; set; } = "";
    // Автозапуск включается автоматически ОДИН раз (первый запуск после установки).
    // Флаг не даёт включить его снова, если пользователь снял тумблер в Settings.
    [JsonPropertyName("autostart_configured")] public bool                    AutostartConfigured { get; set; }
    // Оверлейные правки плейлистов из iOS (Фаза 0). Подмешиваются в /api/data на
    // чтении; в master.db на Windows не уезжают (записи ещё нет).
    [JsonPropertyName("playlists")]         public List<PlaylistOverlay>      Playlists        { get; set; } = new();
}

public sealed class DataStore
{
    public static readonly DataStore Shared = new();

    private readonly object _lock = new();
    private RimoData _data = new();

    public RimoData Data { get { lock (_lock) return _data; } }

    private DataStore()
    {
        _data = Load();
    }

    private RimoData Load()
    {
        try
        {
            if (!File.Exists(AppConfig.Shared.DataFile)) return new RimoData();
            var json = File.ReadAllText(AppConfig.Shared.DataFile);
            return JsonSerializer.Deserialize<RimoData>(json) ?? new RimoData();
        }
        catch { return new RimoData(); }
    }

    public void Save(RimoData data)
    {
        lock (_lock) { _data = data; }
        Task.Run(() =>
        {
            try
            {
                var json = JsonSerializer.Serialize(data, new JsonSerializerOptions { WriteIndented = true });
                // Atomic write: a crash/shutdown mid-write must not leave a truncated
                // JSON that throws on the next Load() and resets the store (→ unpair).
                // System.Text.Json already tolerates missing keys (new fields keep
                // their C# defaults), so schema evolution alone is safe here; the temp
                // + Move guards the corrupt-file path.
                var path = AppConfig.Shared.DataFile;
                var tmp  = path + ".tmp";
                File.WriteAllText(tmp, json);
                File.Move(tmp, path, overwrite: true);
            }
            catch (Exception ex) { Log.Error($"DataStore save failed: {ex.Message}"); }
        });
    }

    public void Update(Action<RimoData> action)
    {
        RimoData copy;
        lock (_lock) { copy = JsonSerializer.Deserialize<RimoData>(JsonSerializer.Serialize(_data))!; }
        action(copy);
        Save(copy);
    }

    /// Generate the per-device LAN PSK at startup if none exists yet, so the
    /// agent's own WinUI control calls — which authenticate over 127.0.0.1 with
    /// ?lan_token= — work before the pairing QR has ever been shown (task 6004).
    /// Idempotent: a no-op once a secret is present.
    public void EnsureLanSecret()
    {
        if (!string.IsNullOrEmpty(Data.LanSecret)) return;
        var raw = new byte[32];
        System.Security.Cryptography.RandomNumberGenerator.Fill(raw);
        var psk = Convert.ToBase64String(raw).Replace("+", "-").Replace("/", "_").Replace("=", "");
        Update(d => { if (string.IsNullOrEmpty(d.LanSecret)) d.LanSecret = psk; });
    }

    /// Оверлеи, которые ещё отличаются от того, что записано в master.db (Фаза 6).
    /// Единственный источник правды и для движка синка, и для счётчика pending_sync
    /// в /api/data — они не имеют права разойтись, иначе кнопка Sync либо предлагает
    /// пустой прогон, либо прячет нужный.
    ///
    /// Tombstone плейлиста, который до базы так и не доехал, удалять там нечего —
    /// он НЕ pending. LastSyncedHash == null ⇒ ни разу не синкали ⇒ pending.
    public List<PlaylistOverlay> PendingSyncOverlays()
    {
        return Data.Playlists
            .Where(ov => !(ov.Deleted && string.IsNullOrEmpty(ov.RekordboxId))
                         && ov.LastSyncedHash != ov.SyncSignature)
            .ToList();
    }
}
